src/Writer/XlsxWriter.php

```diff
@@ -0,0 +1,222 @@
+     */
+    private $spreadsheet;
+
+    /**
+     * @var Worksheet
+     */
+    private $worksheet;
+
+    /**
+     * @var int
+     */
+    private $position;
+
+    /**
+     * @throws \LogicException
+     * @throws \RuntimeException
+     */
+    public function __construct(string $filename, bool $showHeaders = true, bool $showFilters = true)
+    {
+        $this->filename = $filename;
+        $this->showHeaders = $showHeaders;
+        $this->showFilters = $showFilters;
+        $this->position = 1;
+
+        if (!class_exists(Spreadsheet::class)) {
+            throw new \LogicException('You need the "phpoffice/spreadsheet" component in order to use the XLSX export.');
+        }
+
+        if (is_file($filename)) {
+            throw new \RuntimeException(sprintf('The file "%s" already exist', $filename));
+        }
+    }
+
+    public function getDefaultMimeType(): string
+    {
+        return 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet';
+    }
+
+    public function getFormat(): string
+    {
+        return 'xlsx';
+    }
+
+    public function open(): void
+    {
+        $this->spreadsheet = new Spreadsheet();
+
+        $this->worksheet = $this->spreadsheet->getActiveSheet();
+    }
+
+    /**
+     * @throws WriterException
+     */
+    public function close(): void
+    {
+        if ($this->showHeaders && $this->showFilters) {
+            $this->worksheet->setAutoFilter($this->worksheet->calculateWorksheetDimension());
+            $this->worksheet->setSelectedCellByColumnAndRow(1, 1);
+        }
+
+        $excelWriter = IOFactory::createWriter($this->spreadsheet, 'Xlsx');
+        $excelWriter->save($this->filename);
+    }
+
+    public function write(array $data): void
+    {
+        if (1 === $this->position && $this->showHeaders) {
+            $this->addHeaders($data);
+
+            ++$this->position;
+        }
+
+        $column = 1;
+
+        foreach ($data as $value) {
+            $dataFormat = $this->getDataFormat($value);
+            $dataValue = $this->getDataValue($value);
+
+            if (null !== $dataFormat) {
+                $this->worksheet->getStyleByColumnAndRow($column, $this->position)
+                    ->getNumberFormat()
+                    ->setFormatCode($dataFormat);
+            }
+
+            $dataType = $this->getDataType($value);
+
+            $this->worksheet->setCellValueExplicitByColumnAndRow($column, $this->position, $dataValue, $dataType);
+
+            ++$column;
+        }
+
+        ++$this->position;
+    }
+
+    private function addHeaders(array $data): void
+    {
+        $column = 1;
+
+        foreach (array_keys($data) as $value) {
+            $this->worksheet->setCellValueExplicitByColumnAndRow($column, $this->position, $value, DataType::TYPE_STRING);
+
+            ++$column;
+        }
+    }
+
+    /**
+     * Get the type of the Spreadsheet cell.
+     */
+    private function getDataType($value): string
```

Comment:
can we add `mixed` as parameter type in phpdoc? (this applies to other `private` functions)

Reply:
Done